File: 2wordladders/WordLadders.java

import java.util.*;

public class WordLadders {
    private ArrayList<Pair> pairs;
    private ArrayList<String> words;
    private HashMap<String, LinkedList<String>> neighborList = new HashMap<>();

    public static void main(String[] args) {
        new WordLadders().run(args);
    }

    public void run(String[] args) {
        Parser parser = new Parser(new Scanner(System.in));
        this.pairs = parser.getPairs();
        this.words = parser.getWords();

        // creates a list of neighbour for every word we have
        createNeighborList();

        // Calculate shortest path between two words
        for (Pair p : pairs) {
            int res = bfs(p.s1, p.s2);
            if (res == -1) {
                System.out.println("Impossible");
            } else {
                System.out.println(res);
            }
        }
    }

    private List<String> getNeighborList(String s) {
        char[] array = s.substring(1).toCharArray();
        Arrays.sort(array);
        return neighborList.get(String.valueOf(array));
    }

    private void createNeighborList() {
        for (String s : words) {
            /* by sorting them we can compare strings easier */
            char[] array = s.toCharArray();
            Arrays.sort(array);
            String sortedArray = String.valueOf(array);
            for (int i = 0; i < 5; i++) {
                // draw an arc only if the last four letters are present
                String rep = sortedArray.substring(0, i) + sortedArray.substring(i + 1);
                /*
                 * inptu
                 * i = 0: = nptu
                 * i = 1: i + ptu = iptu
                 * i = 2: in + tu = intu
                 * i = 3: inp + u = inpt
                 * i = 4: = inpt
                 */
                // Create new list for each new neighbor node, if there isn't already.
                neighborList.computeIfAbsent(rep, e -> new LinkedList<>());
                // retrieve the list and add the word as neighbor.
                List<String> neighbours = neighborList.get(rep);
                if (neighbours != null) {
                    neighbours.add(s);
                }
            }
        }
    }

    private int getSize(HashMap<String, String> pred, String t) {
        int counter = -1;
        if (pred.get(t) != null) {
            //start at string t and loop back each node to count the size of graph.
            for (String str = t; str != null; str = pred.get(str)) {
                counter++;
            }
        }
        return counter;
    }

    public int bfs(String s, String t) {
        HashMap<String, String> pred = new HashMap<>();
        Map<String, Boolean> visited = new HashMap<>();
        Queue<String> q = new LinkedList<>(); // new list q containing node s, the first one we visit
        q.add(s); // we are visiting that node first
        visited.put(s, true); // first node we visit
        if (s.equals(t)) { // return if already at destination
            return 0;
        }
        while (!q.isEmpty()) {
            String v = q.poll(); // take out the first element form q
            for (String w : getNeighborList(v)) {
                if (visited.get(w) == null) { // if the neighbor string is not visited yet
                    visited.put(w, true);
                    q.add(w); // we go to that node by adding to the queue
                    pred.put(w, v); // now this new node and previous node are connected
                    if (w.equals(t)) { // if that node is the string t we are looking for   
                        return getSize(pred, t);
                    }
                }
            }
        }
        return -1;
    }
}

class Pair {
    String s1;
    String s2;

    public Pair(String s1, String s2) {
        this.s1 = s1;
        this.s2 = s2;
    }
}

class Parser {
    private Scanner scan;
    private ArrayList<Pair> pairs;
    private ArrayList<String> words;

    public Parser(Scanner scan) {
        this.scan = scan;
        this.pairs = new ArrayList<>();
        this.words = new ArrayList<>();
        read();
    }

    private void read() {
        // scan first two integers
        int numberOfWords = scan.nextInt();
        int numberOfQueries = scan.nextInt();
        // scan the words
        for (var i = 0; i < numberOfWords; i++) {
            words.add(scan.next());
        }
        // Scan pairs
        for (var i = 0; i < numberOfQueries; i++) {
            String firstWord = scan.next();
            String secondWord = scan.next();
            pairs.add(new Pair(firstWord, secondWord));
        }
        scan.close();
    }

    public ArrayList<Pair> getPairs() {
        return pairs;
    }

    public ArrayList<String> getWords() {
        return words;
    }
}

File: 2wordladders/WordLadders.py
from collections import deque


class WordLadders:
    def __init__(self):
        self.pairs = []
        self.words = []
        self.parser()
        self.neighborList = {}
        self.createNeighborList()
        self.run()

    def parser(self):
        firstLine = input().strip().split()
        nbrOfWords = int(firstLine[0])
        nbrOfQueries = int(firstLine[1])
        for _ in range(nbrOfWords):
            self.words.append(input())
        for _ in range(nbrOfQueries):
            line = input().strip().split()
            first_word = line[0]
            second_word = line[1]
            self.pairs.append((first_word, second_word))

    def run(self):
        for p1, p2 in self.pairs:
            res = self.bfs(p1, p2)
            if res == -1:
                print("Impossible")
            else:
                print(res)

    def createNeighborList(self):
        for word in self.words:
            char_array = sorted(word)
            sorted_string = "".join(char_array)
            for i in range(5):
                rep = sorted_string[:i] + sorted_string[i + 1 :]
                self.neighborList.setdefault(rep, []).append(word)

    def getNeighborList(self, s):
        array = sorted(s[1:])
        sorted_array = "".join(array)
        return self.neighborList.get(sorted_array)

    def bfs(self, start, end):
        if start == end:
            return 0

        pred = dict()
        visited = set()
        visited.add(start)
        queue = deque([start])

        while queue:
            v = queue.popleft()
            for w in self.getNeighborList(v):
                if w not in visited:
                    visited.add(w)
                    queue.append(w)
                    pred[w] = v
                    if w == end:
                        return self.getSize(pred, end)
        return -1

    def getSize(self, pred, t):
        counter = -1
        curr = t
        while curr is not None:
            counter += 1
            if pred.get(curr) is None:
                break
            curr = pred[curr]
        return counter


if __name__ == "__main__":
    WordLadders()